exp/kubeadm-ignition/api/v1alpha3/kubeadmbootstrapconfig_types.go

```diff
@@ -0,0 +1,337 @@
+	// +optional
+	Mounts []MountPoints `json:"mounts,omitempty"`
+
+	// PreKubeadmCommands specifies extra commands to run before kubeadm runs
+	// +optional
+	PreKubeadmCommands []string `json:"preKubeadmCommands,omitempty"`
+
+	// PostKubeadmCommands specifies extra commands to run after kubeadm runs
+	// +optional
+	PostKubeadmCommands []string `json:"postKubeadmCommands,omitempty"`
+
+	// Users specifies extra users to add
+	// +optional
+	Users []User `json:"users,omitempty"`
+
+	// NTP specifies NTP configuration
+	// +optional
+	NTP *NTP `json:"ntp,omitempty"`
+
+	// Format specifies the output format of the bootstrap data
+	// +optional
+	Format Format `json:"format,omitempty"`
+
+	// Verbosity is the number for the kubeadm log level verbosity.
+	// It overrides the `--v` flag in kubeadm commands.
+	// +optional
+	Verbosity *int32 `json:"verbosity,omitempty"`
+
+	// UseExperimentalRetryJoin replaces a basic kubeadm command with a shell
+	// script with retries for joins.
+	//
+	// This is meant to be an experimental temporary workaround on some environments
+	// where joins fail due to timing (and other issues). The long term goal is to add retries to
+	// kubeadm proper and use that functionality.
+	//
+	// This will add about 40KB to userdata
+	//
+	// For more information, refer to https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055.
+	// +optional
+	UseExperimentalRetryJoin bool `json:"useExperimentalRetryJoin,omitempty"`
+}
+
+// KubeadmConfigStatus defines the observed state of KubeadmConfig
+type KubeadmConfigStatus struct {
+	// Ready indicates the BootstrapData field is ready to be consumed
+	Ready bool `json:"ready,omitempty"`
+
+	// DataSecretName is the name of the secret that stores the bootstrap data script.
+	// +optional
+	DataSecretName *string `json:"dataSecretName,omitempty"`
+
+	// BootstrapData will be a cloud-init script for now.
+	//
+	// Deprecated: This field has been deprecated in v1alpha3 and
+	// will be removed in a future version. Switch to DataSecretName.
+	//
+	// +optional
+	BootstrapData []byte `json:"bootstrapData,omitempty"`
+
+	// FailureReason will be set on non-retryable errors
+	// +optional
+	FailureReason string `json:"failureReason,omitempty"`
+
+	// FailureMessage will be set on non-retryable errors
+	// +optional
+	FailureMessage string `json:"failureMessage,omitempty"`
+
+	// ObservedGeneration is the latest generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Conditions defines current service state of the KubeadmConfig.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kubeadmconfigs,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+// KubeadmConfig is the Schema for the kubeadmconfigs API
+type KubeadmConfig struct {
```

Comment:
I suspect this type will need to be renamed to avoid conflicting with the cloud-init based bootstrap provider.

Reply:
> I suspect this type will need to be renamed to avoid conflicting with the cloud-init based bootstrap provider.

Right. I renamed related structures to `KubeadmIgnition*`.